panda-rs/src/api/regs.rs

```diff
@@ -0,0 +1,212 @@
+use crate::prelude::*;
+use crate::cpu_arch_state;
+
+use strum_macros::{EnumString, EnumIter};
+use strum::IntoEnumIterator;
+
+// Arch-specific mappings ----------------------------------------------------------------------------------------------
+
+// TODO: handle AX/AH/AL, etc via shifts? Tricky b/c enum val used to index QEMU array
+#[cfg(feature = "i386")]
+#[derive(Debug, Copy, Clone, PartialEq, Eq, EnumString, EnumIter)]
+pub enum Reg {
+    EAX = 0,
+    ECX = 1,
+    EDX = 2,
+    EBX = 3,
+    ESP = 4,
+    EBP = 5,
+    ESI = 6,
+    EDI = 7,
+}
+
+// TODO: handle EAX/AX/AH/AL, etc via shifts? Tricky b/c enum val used to index QEMU array
+#[cfg(feature = "x86_64")]
+#[derive(Debug, Copy, Clone, PartialEq, Eq, EnumString, EnumIter)]
+pub enum Reg {
+    RAX = 0,
+    RCX = 1,
+    RDX = 2,
+    RBX = 3,
+    RSP = 4,
+    RBP = 5,
+    RSI = 6,
+    RDI = 7,
+    R8 = 8,
+    R9 = 9,
+    R10 = 10,
+    R11 = 11,
+    R12 = 12,
+    R13 = 13,
+    R14 = 14,
+    R15 = 15,
+}
+
+#[cfg(feature = "arm")]
+#[derive(Debug, Copy, Clone, PartialEq, Eq, EnumString, EnumIter)]
+pub enum Reg {
+    R0 = 0,
+    R1 = 1,
+    R2 = 2,
+    R3 = 3,
+    R4 = 4,
+    R5 = 5,
+    R6 = 6,
+    R7 = 7,
+    R8 = 8,
+    R9 = 9,
+    R10 = 10,
+    R11 = 11,
+    R12 = 12,
+    LR = 13,
+    SP = 14,
+    IP = 15,
+}
+
+#[cfg(any(feature = "mips", feature = "mipsel"))]
+#[derive(Debug, Copy, Clone, PartialEq, Eq, EnumString, EnumIter)]
+pub enum Reg {
+    ZERO = 0,
+    AT = 1,
+    V0 = 2,
+    V1 = 3,
+    A0 = 4,
+    A1 = 5,
+    A2 = 6,
+    A3 = 7,
+    T0 = 8,
+    T1 = 9,
+    T2 = 10,
+    T3 = 11,
+    T4 = 12,
+    T5 = 13,
+    T6 = 14,
+    T7 = 15,
+    S0 = 16,
+    S1 = 17,
+    S2 = 18,
+    S3 = 19,
+    S4 = 20,
+    S5 = 21,
+    S6 = 22,
+    S7 = 23,
+    T8 = 24,
+    T9 = 25,
+    K0 = 26,
+    K1 = 27,
+    GP = 28,
+    SP = 29,
+    FP = 30,
+    RA = 31,
+}
+
+// TODO: reg map
+//#[cfg(feature = "aarch64")]
+//#[derive(Debug, PartialEq, Eq, EnumString, EnumIter)]
+
+// TODO: reg map
+//#[cfg(feature = "ppc")]
+//#[derive(Debug, PartialEq, Eq, EnumString, EnumIter)]
+
+// Getters/setters -----------------------------------------------------------------------------------------------------
+
+/// Get stack pointer register
+pub fn reg_sp() -> Reg {
+
+    #[cfg(feature = "i386")]
+    return Reg::ESP;
+
+    #[cfg(feature = "x86_64")]
+    return Reg::RSP;
+
+    #[cfg(any(feature = "arm", feature = "mips", feature = "mipsel"))]
+    return Reg::SP;
+}
+
+/// Get return value register
+#[cfg(any(feature = "i386", feature = "x86_64"))]
+pub fn reg_ret_val() -> Reg {
+
+    #[cfg(feature = "i386")]
+    return Reg::EAX;
+
+    #[cfg(feature = "x86_64")]
+    return Reg::RAX;
+}
+
+/// Get return value registers
+/// Note that most C code will only use the first register, e.g. index 0 in returned `Vec`
+#[cfg(any(feature = "arm", feature = "mips", feature = "mipsel"))]
+pub fn reg_ret_val() -> Vec<Reg> {
```

Comment:
is there a reason for this being a `Vec<Reg>` instead of, say, an `&'static [Reg]`?

(Also maybe some of these should be constants, I'm curious your thoughts on the matter)

Reply:
Nope! Will switch to the latter to avoid the alloc. Also, looking back at it, will merge the two variants of this function so that there's a single function with a `&'static [Reg]` return (slice with one element for arches where a single reg is used) - figure this makes the API for writing plugins consistent regardless of target arch.

Having functions mirrors the Python API, but there's no requirement to do so - what about `const fn`? Best of both worlds?

Reply:
Actually `const fn` wouldn't make sense here since there is no input to these functions! Constants would work, but opting for functions for now just to keep the API a little closer to the Python bindings.